File: Fox.cpp
#include "CONSTANTS.h"
#include "Fox.h"
#include "World.h"

#include <vector>

Fox::Fox(World &world, const Point &position) : Animal(world, position, FOX_SYMBOL, FOX_STRENGTH, FOX_INITIATIVE, FOX_COLOR, FOX_NAME) {}

Organism* Fox::Clone(const Point &position) {
  return new Fox(this->world, position);
}

void Fox::Action() {
  if(CanAction()) {
    std::vector<Point> possibleMoves = this->world.GetPositionsAround(this->GetPosition());
    Point newPosition = this->GetPosition();
    for(auto &move : possibleMoves) {
      if(this->world.GetOrganismAt(move) == nullptr || this->world.GetOrganismAt(move)->GetStrength() <= this->GetStrength()) {
        newPosition = move;
        break;
      }
    }
    if(newPosition != this->GetPosition()) {
      Organism* attackerOrganism = this->world.GetOrganismAt(newPosition);
      if(attackerOrganism == nullptr) { // Free space
        this->world.MoveOrganism(this, newPosition);
      } else { // Collision
        attackerOrganism->SetCanAction(false);
        if(this->GetSymbol() == attackerOrganism->GetSymbol()) {
          this->Collision(attackerOrganism); // Breed
        } else {
          if(this->GetStrength() >= attackerOrganism->GetStrength()) {
            bool takenField = attackerOrganism->Collision(this); // Kill
            if(!takenField && attackerOrganism->GetSymbol() != HOGWEED_SYMBOL && attackerOrganism->GetSymbol() != NIGHTSHADE_BERRIES_SYMBOL) {
              this->world.MoveOrganism(this, newPosition);
            }
          } else {
            if(attackerOrganism->GetSymbol() == HOGWEED_SYMBOL || attackerOrganism->GetSymbol() == NIGHTSHADE_BERRIES_SYMBOL) { // Hogweed and NightshadeBerries kill the attacker, but they're also eaten
              attackerOrganism->Collision(this);
            } else {
              this->Collision(attackerOrganism); // Killed
            }
          }
        }
      }
    }
  }
}

File: Java/Fox.java
import java.util.Vector;
import java.util.Collections;
public class Fox extends Animal {
  public Fox(World world, Point position) {
    super(world, position, Constants.FOX_SYMBOL, Constants.FOX_STRENGTH, Constants.FOX_INITIATIVE, Constants.FOX_COLOR, Constants.FOX_NAME, Constants.ANIMAL_RANGE);
  }

  @Override
  public final Organism Clone(Point position) {
    return new Fox(this.world, position);
  }

  @Override
  public final void Action() {
    if (CanAction()) {
      Vector<Point> possibleMoves = this.world.GetPositionsAround(this.GetPosition());
      Collections.shuffle(possibleMoves);
      Point newPosition = null;
      for (Point move : possibleMoves) {
        Organism other = world.GetOrganismAt(move);
        if (other == null || other.GetStrength() < this.GetStrength()) {
          newPosition = move;
          break;
        }
      }
      if (newPosition != null) {
        Organism other = world.GetOrganismAt(newPosition);
        if (other == null) {
          this.world.MoveOrganism(this, newPosition);
        } else { // Collision
          other.SetCanAction(false);
          if (this.GetSymbol() == other.GetSymbol()) {
            this.Collision(other); // Breed
          } else {
            if (this.GetStrength() >= other.GetStrength()) {
              boolean takenField = other.Collision(this); // Kill
              if (!takenField && other.GetSymbol() != Constants.HOGWEED_SYMBOL && other.GetSymbol() != Constants.NIGHTSHADE_BERRIES_SYMBOL) {
                this.world.MoveOrganism(this, newPosition);
              }
            } else {
              if (other.GetSymbol() == Constants.HOGWEED_SYMBOL || other.GetSymbol() == Constants.NIGHTSHADE_BERRIES_SYMBOL) { // Hogweed and Nightshade Berries kill the attacker, but they are also eaten
                other.Collision(this);
              } else {
                this.Collision(other); // Killed
              }
            }
          }
        }
      }
    }
  }
}
